core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -630,44 +703,73 @@ class LogManager(logDirs: Array[File],
   }
 
   /**
+    * Mark the partition directory in the source log directory for deletion and
+    * rename the future log of this partition in the destination log directory to be the current log
+    *
+    * @param topicPartition TopicPartition that needs to be swapped
+    */
+  def replaceCurrentWithFutureLog(topicPartition: TopicPartition): Unit = {
+    logCreationOrDeletionLock synchronized {
+      val sourceLog = currentLogs.get(topicPartition)
+      val destLog = futureLogs.get(topicPartition)
+
+      if (sourceLog == null)
+        throw new KafkaStorageException(s"The current replica for $topicPartition in ${sourceLog.dir.getParent} is offline")
+      if (destLog == null)
+        throw new KafkaStorageException(s"The future replica for $topicPartition in ${destLog.dir.getParent} is offline")
+
+      destLog.renameDir(Log.logDirName(topicPartition))
+      // Now that future replica has been successfully renamed to be the current replica
+      // Update the cached map and log cleaner as appropriate.
+      futureLogs.remove(topicPartition)
+      currentLogs.put(topicPartition, destLog)
+      if (cleaner != null) {
+        cleaner.alterCheckpointDir(topicPartition, sourceLog.dir.getParentFile, destLog.dir.getParentFile)
+        cleaner.resumeCleaning(topicPartition)
+      }
+
+      // Close handlers of log files in the source log directory in advance before renaming source log for deletion
+      sourceLog.closeHandlers()
```

Comment:
I think we need to call sourceLog.closeHandlers() instead of sourceLog.close(). If we call sourceLog.close() and if source log directory is offline, this method will throw KafkaStorageException without closing handler for this sourceLog. Later when LogManager.handleLogDirFailure() is called for this source log directory, the handles of this sourceLog will not be closed because sourceLog has already been removed from `currentLogs`.

On the other hand, if we sourceLog.closeHandlers() and if the source log directory is offline, the handler of sourceLog will be closed before sourceLog.renameDir() throws KafkaStorageException. All handlers in the source log directory will be properly closed in this case.

Does this make sense?